File: leet/min_cost_for_tickets/min_cost_for_tickets.java
// LeetCode #983. Minimum Cost For Tickets
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        Set<Integer> daySet = new HashSet<Integer>(Arrays.stream(days).boxed().collect(Collectors.toList()));
        int[] dp = new int[366];
        for (int i = 1; i < 366; i++)
        {
            if (!daySet.contains(i))
                dp[i] = dp[i - 1];
            else
            {
                dp[i] = dp[i - 1] + costs[0];
                dp[i] = Math.min(dp[i], dp[Math.max(0, i - 7)] + costs[1]);
                dp[i] = Math.min(dp[i], dp[Math.max(0, i - 30)] + costs[2]);
            }
        }
        return dp[365];
    }
}

File: leet/min_cost_for_tickets/min_cost_for_tickets.py
# LeetCode #983. Minimum Cost For Tickets
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0] * 366
        days_set = set(days)
        for i in range(1, len(dp)):
            if i not in days_set:
                dp[i] = dp[i - 1]
                continue
            dp[i] = dp[i - 1] + costs[0]
            dp[i] = min(dp[i], dp[i - 7] + costs[1])
            dp[i] = min(dp[i], dp[i - 30] + costs[2])
        return dp[-1]
